### ISSUE TITLE
Hatching in legends is broken

### ISSUE THREAD

**Issue by tacaswell** [matplotlib/matplotlib#7166]
Compare http://matplotlib.org/2.0.0b4/examples/api/filled_step.html (2.0.0b4)

![image](https://cloud.githubusercontent.com/assets/199813/18773638/d820ea30-811e-11e6-9aff-d839eabeea00.png)

and 

http://matplotlib.org/examples/api/filled_step.html (1.5.3)

![image](https://cloud.githubusercontent.com/assets/199813/18773629/ce9d0fc0-811e-11e6-9aef-054cf0f8d570.png)

**Comment by Kojoley** [comment]
2.0.0b3 is fine
http://matplotlib.org/2.0.0b3/examples/api/filled_step.html

**Comment by WeatherGod** [comment]
Also, legends are used in the cycle tests, which includes cycling hatches
for ax.fill() calls. So, that can't be broken or Travis would have noticed
it. So this must be very specific to this plotting function.

**Comment by Kojoley** [comment]
Broken by #6904

**Comment by Kojoley** [comment]
Do not we have a test for legends hatching? I thought we have...

**Comment by WeatherGod** [comment]
Well, like I said, it does work for ax.fill(), but the bug report is for
filled-step plots, which are different enough, it seems.

**Comment by Kojoley** [comment]
9982e400c7b21afdebe364d8eb1a8565c5383242
![filled_step-9982e400c7b21afdebe364d8eb1a8565c5383242](https://cloud.githubusercontent.com/assets/2743474/18793048/c7202518-81c1-11e6-9fe7-ec68ca3f6647.png)

7bf436c426cfc4f7e2766ff0622ebbdd52d7f6ff
![filled_step-7bf436c426cfc4f7e2766ff0622ebbdd52d7f6ff](https://cloud.githubusercontent.com/assets/2743474/18792962/52b7bcf4-81c1-11e6-9e25-cfee8141cbc7.png)

diff
![filled_step-7bf436c426cfc4f7e2766ff0622ebbdd52d7f6ff-failed-diff](https://cloud.githubusercontent.com/assets/2743474/18793002/845aa910-81c1-11e6-8124-7178fa39d702.png)

* legend frame was disabled for bisecting

**Comment by WeatherGod** [comment]
strange... that commit didn't do anything with hatches as far as I can
tell. But it is a major rework of patch objects, so maybe something got
missed?

Also, that doesn't explain how everything works fine for ax.fill() and
legend() (from the cycler tests).

**Comment by WeatherGod** [comment]
Waitaminute... could this be a polygon vs. patch issue?

**Comment by Kojoley** [comment]
> Waitaminute... could this be a polygon vs. patch issue?

I am thinking about the same.

**Comment by afvincent** [comment]
Seems to be related: the legend handles of the example [contourf_hatching](http://matplotlib.org/devdocs/examples/pylab_examples/contourf_hatching.html) appear to be missing (vs the same [example from the 1.5.3 gallery](http://matplotlib.org/examples/pylab_examples/contourf_hatching.html))

**Comment by dopplershift** [comment]
So it's not about the hatching setting (that's getting put in correctly into the handles), but there's something not right about setting the color used for the hatching; even passing `alpha=1.0` (which should be a noop) is enough to get the proper result.  This explains why 7bf436c426cfc4f7e2766ff0622ebbdd52d7f6ff is the offending commit.

**Comment by dopplershift** [comment]
And setting rcParam `patch.force_edgecolor` to `True` fixes it as well.

**Comment by dopplershift** [comment]
I may be getting out of my depth here. This example reproduces simply:
```python
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import matplotlib.collections as mcoll

fig, ax = plt.subplots(1, 1)
rect = mpatches.Rectangle((0, 0), 3, 4, hatch='/')
ax.add_patch(rect)
rect2 = mcoll.RegularPolyCollection(4, sizes=[20000], offsets=[(5, 5)],
                                    transOffset=ax.transData, hatch='/')
ax.add_collection(rect2)
ax.set_xlim(0, 7)
ax.set_ylim(0, 9)
```
On matplotlib 1.5.3 this gives:
![image](https://cloud.githubusercontent.com/assets/221526/19873914/43ef1d6e-9f87-11e6-8928-cc6d0088b70c.png)

On 2.x:
![image](https://cloud.githubusercontent.com/assets/221526/19873928/4e8f9d02-9f87-11e6-938d-82776fce0e88.png)

Apparently there's some way that the collection accesses/sets the edgecolor for hatching (and not the outline) that patch is missing. I can't seem to find it though.

**Comment by QuLogic** [comment]
Thanks for the example @dopplershift; it made it (relatively) easy to figure out what is going on. I believe the following explains the difference:

For collections:
  1. The default edge colour (without forcing) is [`'none'`](https://github.com/matplotlib/matplotlib/blob/07adcdeb367c008229daafa932397c735123cd2d/lib/matplotlib/collections.py#L687), which after going through [`mcolors.to_rgba_array`](https://github.com/matplotlib/matplotlib/blob/07adcdeb367c008229daafa932397c735123cd2d/lib/matplotlib/collections.py#L701) is an empty array `[]`.
 2. When hatching is enabled, collections use [`renderer.draw_path_collection`](https://github.com/matplotlib/matplotlib/blob/07adcdeb367c008229daafa932397c735123cd2d/lib/matplotlib/collections.py#L338) to draw.
 3. This [function](https://github.com/matplotlib/matplotlib/blob/07adcdeb367c008229daafa932397c735123cd2d/src/_backend_agg.h#L903) a) will set a default [`gc.linewidth = 0.0`](https://github.com/matplotlib/matplotlib/blob/07adcdeb367c008229daafa932397c735123cd2d/src/_backend_agg.h#L950) and b) [does not modify `gc.color` when no edge colours are specified](https://github.com/matplotlib/matplotlib/blob/07adcdeb367c008229daafa932397c735123cd2d/src/_backend_agg.h#L991-L1003).
 4. It dispatches to [`_draw_path`](https://github.com/matplotlib/matplotlib/blob/07adcdeb367c008229daafa932397c735123cd2d/src/_backend_agg.h#L307) for each path in the collection, which a) draws the [hatching with `gc.color`](https://github.com/matplotlib/matplotlib/blob/07adcdeb367c008229daafa932397c735123cd2d/src/_backend_agg.h#L374) at `gc.hatch_linewidth` and b) draws [no stroke when `gc.linewidth == 0.0`](https://github.com/matplotlib/matplotlib/blob/07adcdeb367c008229daafa932397c735123cd2d/src/_backend_agg.h#L408).

The default `gc.color` is [opaque black](https://github.com/matplotlib/matplotlib/blob/07adcdeb367c008229daafa932397c735123cd2d/lib/matplotlib/backend_bases.py#L790), so the hatch is drawn in black, while the edge is not stroked because of 0 line width.

On the other hand, for patches:
 1. The default edge colour is [`'none'`](https://github.com/matplotlib/matplotlib/blob/07adcdeb367c008229daafa932397c735123cd2d/lib/matplotlib/patches.py#L294) again, but is passed through [`mcolors.to_rgba` instead](https://github.com/matplotlib/matplotlib/blob/07adcdeb367c008229daafa932397c735123cd2d/lib/matplotlib/patches.py#L295) and the result is fully transparent black.
 2. When drawing, this edge colour is then [translated to `gc.color`](https://github.com/matplotlib/matplotlib/blob/07adcdeb367c008229daafa932397c735123cd2d/lib/matplotlib/patches.py#L523); also, [`gc.linewidth` is set to 0 if the colour is transparent](https://github.com/matplotlib/matplotlib/blob/07adcdeb367c008229daafa932397c735123cd2d/lib/matplotlib/patches.py#L527).
 3. Eventually `_draw_path` is called again, but this time the hatch is drawn in transparent black.

So I guess this basically stems from hatching using the edge colour, but not its line width, making it only semi-independent. If we a) went back to using line width for invisible edges, or b) fully separated hatching styling from edge styling, then maybe this would not occur. I'm not sure if there's some way to get a middle ground. The reason that collections work is sort of a happy accident of a few things.

**Comment by efiring** [comment]
Logically, the hatching style parameters should be independent of the edge parameters, so I think that your option b is the one to pursue.

**Comment by dopplershift** [comment]
I agree with @efiring on making hatching its own thing. It wasn't clear to me until I dug in that edge color controls hatching color.

Yet more weirdness in matplotlib that only reveals itself when we fiddle with defaults.

**Comment by dopplershift** [comment]
Fixed by #7415.
